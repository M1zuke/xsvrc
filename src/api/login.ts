import { isErrorType } from '../store/reducer';
import { setUserInfo } from '../store/user/actions';
import { AppThunkAction } from '../thunk';
import { getAllFavoriteGroups, getAllFavorites } from './favorites-api';
import { getAllFriends } from './friends-api';
import { logout } from './logout';
import { getAllModerations } from './moderations';
import { getAllNotifications } from './notifications';
import { api, prepare } from './prepare';
import { AuthenticatedUserInfo } from './types';

export function login(username?: string, password?: string): AppThunkAction<Promise<void>> {
  return async function (dispatch, getState) {
    const state = getState();
    if (state.user.userInfo === null || isErrorType(state.user.userInfo)) {
      dispatch(setUserInfo('loading'));

      const authorizationBuffer = Buffer.from(`${username}:${password}`);
      const base64 = authorizationBuffer.toString('base64');

      const result = await prepare<AuthenticatedUserInfo>(state, dispatch, {
        url: api('auth/user'),
        headers: {
          Authorization: `Basic ${base64}`,
        },
      });

      if (result.type === 'entity') {
        dispatch(setUserInfo(result.result));
        await dispatch(getAllFriends()).finally();
        await dispatch(getAllNotifications()).finally();
        await dispatch(getAllFavoriteGroups()).finally();
        await dispatch(getAllFavorites()).finally();
        dispatch(getAllModerations()).finally();
      } else {
        dispatch(logout()).finally();
      }
    }
  };
}
