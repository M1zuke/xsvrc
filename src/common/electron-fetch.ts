import { IpcRenderer } from '../index';

export type ElectronFetchArgs

export function electronFetch<T>(): Promise<T> {
  return IpcRenderer.invoke('fetch', (e,args: ElectronFetchArgs) => {

  })
}
