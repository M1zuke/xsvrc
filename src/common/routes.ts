export type RouteDefinition = {
  config: {
    path: string;
    exact?: boolean;
  };
  path: string;
};

export type Routes = {
  home: RouteDefinition;
  login: RouteDefinition;
  friends: RouteDefinition;
  eventList: RouteDefinition;
  notifications: RouteDefinition;
  settings: RouteDefinition;
  moderation: RouteDefinition;
  avatar: RouteDefinition;
  userSearch: RouteDefinition;
  transactions: RouteDefinition;
};

export const routes: Routes = {
  home: { path: '/', config: { path: '/', exact: true } },
  login: { path: '/login', config: { path: '/login', exact: true } },
  friends: { path: '/friends', config: { path: '/friends', exact: true } },
  eventList: { path: '/user-event-list', config: { path: '/user-event-list', exact: true } },
  notifications: { path: '/notifications', config: { path: '/notifications', exact: true } },
  settings: { path: '/settings', config: { path: '/settings', exact: true } },
  moderation: { path: '/moderation', config: { path: '/moderation', exact: true } },
  avatar: { path: '/avatar', config: { path: '/avatar', exact: true } },
  userSearch: { path: '/user-search', config: { path: '/user-search', exact: true } },
  transactions: { path: '/transactions', config: { path: '/transactions', exact: true } },
};
