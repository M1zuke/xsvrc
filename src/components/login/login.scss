@import "../../utils/colors";

.login-component {
  background-color: $highlightColor;
  position: relative;
  display: inline-block;
  box-sizing: border-box;
  padding: 20px;
  border-radius: 5px;
  box-shadow: 0 0 3px 1px #000000;
  width: 300px;
  margin: 0 auto;
  top: 30%;

  input {
    display: block;
    background-color: $backgroundColor;
    border: 0;
    padding: 10px 10px;
    border-radius: 0;
    min-width: 200px;
    margin-bottom: 10px;
    color: $textColor;
    width: 100%;
    box-sizing: border-box;
    font-size: 14px;

    &:focus {
      outline-color: $textColor;
    }
  }

  .login-btn {
    margin-top: 10px;
    background-color: $backgroundColor;
    color: $accentColor;
    font-weight: bold;
    display: inline-block;
    padding: 10px 20px;
    cursor: pointer;
    user-select: none;
  }

  .normal-text {
    margin-bottom: 5px;
  }
}
