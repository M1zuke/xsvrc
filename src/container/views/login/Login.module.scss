@import "../../../common/_colors";
@import "../../../common/_spacing";

.Component {
  text-align: center;
  display: grid;
  justify-content: center;
  align-items: center;
  height: 100%;

  .LoginBox {
    background-color: $loginBackgroundColor;
    padding: $loginPadding;
    width: 250px;
    display: grid;
    grid-gap: $loginGridGap;

    .LoginImage {
      height: 100px;
    }
  }
}
