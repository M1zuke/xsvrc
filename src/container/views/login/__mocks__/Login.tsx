import React from 'react';

export const Login = jest.fn(() => <div data-testid="login" />);
