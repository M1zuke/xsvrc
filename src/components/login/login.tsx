import React, { ChangeEvent, FC, useCallback, useState } from 'react';

import './login.scss';
import { useDispatch, useSelector } from 'react-redux';
import { loginUser } from '../../services/vrc-api.service';
import { logInUser } from '../../store/user/actions';
import { selectLoggedIn } from '../../store/user/selectors';
import { Loading } from '../loading/loading';
import { useHistory } from 'react-router';

interface LoginState {
  loggingIn: boolean;
  username: string;
  password: string;
}

export const Login: FC = () => {

  const [state, setState] = useState<LoginState>({
    loggingIn: false,
    username: '',
    password: '',
  });

  const history = useHistory();
  const dispatch = useDispatch();
  const loggedIn: boolean = useSelector(selectLoggedIn);
  const onUsernameChange = useCallback((e: ChangeEvent<HTMLInputElement>) => setState({ ...state, username: e.target.value }), [state]);
  const onPasswordChange = useCallback((e: ChangeEvent<HTMLInputElement>) => setState({ ...state, password: e.target.value }), [state]);


  const onLoginClick = useCallback(() => {
    if (!loggedIn && !state.loggingIn) {
      setState({ ...state, loggingIn: true });
      loginUser(state.username, state.password).then((userInfo: UserInfo) => {
        setState({ ...state, loggingIn: false });
        console.log(userInfo);
        dispatch(logInUser(userInfo));
        history.push('/overview');
      }).catch(() => {
        setState({ ...state, loggingIn: false });
      });
    }
  }, [state, dispatch, history, loggedIn]);

  return (
      <div className="login-component">
        <input className="login-input" type="text" placeholder="Username" onChange={ onUsernameChange }/>
        <input className="password-input" type="password" placeholder="password" onChange={ onPasswordChange }/>
        { state.loggingIn ? <Loading/> : <div className="login-btn" onClick={ onLoginClick }>Login</div> }

      </div>
  );
};
