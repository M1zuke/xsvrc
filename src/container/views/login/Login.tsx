import React, { ReactElement, useCallback, useEffect, useState } from 'react';
import { useSelector } from 'react-redux';
import { useApi } from '../../../api/use-api';
import { Button } from '../../../components/button/Button';
import { TextInput } from '../../../components/input/TextInput';
import { Loading } from '../../../components/loading/Loading';
import { isProbablyAuthenticated } from '../../../components/secured/Secured';
import { useMessages } from '../../../i18n';
import { savedCookies } from '../../../store/cookies/selectors';
import { vrcUserInfo } from '../../../store/user/selectors';
import styles from './Login.module.scss';

export function Login(): ReactElement {
  const messages = useMessages();
  const cookies = useSelector(savedCookies);
  const userInfo = useSelector(vrcUserInfo);
  const { login } = useApi();
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');

  const handleUsernameChange = useCallback((value: string) => setUsername(value), []);
  const handlePasswordChange = useCallback((value: string) => setPassword(value), []);

  const handleLogin = useCallback(() => {
    login(username, password).finally();
  }, [login, password, username]);

  useEffect(() => {
    if (isProbablyAuthenticated(cookies)) {
      handleLogin();
    }
  }, [cookies, handleLogin]);

  return (
    <div className={styles.Component}>
      <div className={styles.LoginBox}>
        <div className={styles.LoginImage} />
        <TextInput
          aria-label="enter username"
          value={username}
          onChange={handleUsernameChange}
          placeholder={messages.Views.Login.View.Username}
        />
        <TextInput
          aria-label="enter password"
          value={password}
          onChange={handlePasswordChange}
          type="password"
          placeholder={messages.Views.Login.View.Password}
        />
        {userInfo === 'loading' ? (
          <Loading />
        ) : (
          <Button aria-label="login" onClick={handleLogin}>
            {messages.Views.Login.View.Login}
          </Button>
        )}
      </div>
    </div>
  );
}
