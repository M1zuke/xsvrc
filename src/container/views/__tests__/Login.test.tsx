import pretty from 'pretty';
import React from 'react';
import { ElectronResult } from '../../../common/electron-fetch';
import * as ElectronFetchModule from '../../../common/electron-fetch';
import { routes } from '../../../common/routes';
import { VRCUserInfo } from '../../../store/user/state';
import { render, withAuthentication } from '../../../test-utils';
import USER_INFO_FETCH_RESULTS from '../../app/__tests__/user-info-fetch-results.json';
import { Login } from '../login/Login';

jest.mock('../../../common/electron-fetch');
const { electronFetch: FetchMock } = ElectronFetchModule as jest.Mocked<typeof ElectronFetchModule>;

describe('<Login /> Component', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders Login correctly', () => {
    const customResult = render(<Login />);
    expect(pretty(customResult.result.container.innerHTML)).toMatchSnapshot();
  });

  it('tries to login if authenticated', () => {
    FetchMock.mockReturnValueOnce(
      new Promise<ElectronResult<VRCUserInfo>>((resolve) => {
        resolve({
          type: 'entity',
          cookies: [],
          result: USER_INFO_FETCH_RESULTS as VRCUserInfo,
        });
      }),
    );

    const customResult = render(<Login />, { state: withAuthentication() });
    expect(customResult.history.location.pathname).toEqual(routes.home.path);
  });
});
