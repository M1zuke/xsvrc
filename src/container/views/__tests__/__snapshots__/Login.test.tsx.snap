// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> Component renders Login correctly 1`] = `
"<div class=\\"Component\\">
  <div class=\\"LoginBox\\">
    <div class=\\"LoginImage\\"></div><input class=\\"Component\\" type=\\"text\\" placeholder=\\"Username\\" value=\\"\\"><input class=\\"Component\\" type=\\"password\\" placeholder=\\"Password\\" value=\\"\\"><button class=\\"Component\\"><span class=\\"StyleElement\\">Login</span></button>
  </div>
</div>"
`;
